src/components/TodoList/TodoList.tsx

```diff
@@ -1,12 +1,55 @@
 /* eslint-disable */
-import React from 'react';
+import React, { useEffect, useState } from 'react';
+import { useDispatch, useSelector } from 'react-redux';
+import { RootState } from '../../app/store';
+import classNames from 'classnames';
+import { setCurrentTodo } from '../../features/currentTodo';
+import { Todo } from '../../types/Todo';
+import { FilterStatus } from '../../features/filter';
 
 export const TodoList: React.FC = () => {
+  const todos = useSelector((state: RootState) => state.todos.todos);
+  const currentTodo = useSelector((state: RootState) => state.currentTodo.todo);
+  const { query, status } = useSelector((state: RootState) => state.filter);
+  const dispatch = useDispatch();
+
+  const filterTodos = (query: string, status: FilterStatus, todos: Todo[]) => {
+    return todos.filter(todo => {
+      const matchesQuery = todo.title
+        .toLowerCase()
+        .includes(query.toLowerCase());
+      const matchesStatus =
+        status === 'all' ||
+        (status === 'active' && !todo.completed) ||
+        (status === 'completed' && todo.completed);
+
+      return matchesQuery && matchesStatus;
+    });
+  };
+
+  const [visibleTodos, setVisibleTodos] = useState(todos);
```

Comment:
So todos, query, and status are stored in state (redux state in this case). Taking it into account - the re-render will happen on any of these variables change. It means there is no reason to store visible todos in state - store them in common variable (and useMemo)